tests/tx/test_tips.py

```diff
@@ -0,0 +1,182 @@
+
+        self.manager.reactor.advance(10)
+
+        self.assertIsNone(self.manager.tx_storage.get_metadata(tx3.hash).voided_by)
+        self.assertIsNotNone(self.manager.tx_storage.get_metadata(tx2.hash).voided_by)
+        # The block confirms tx3, so it's not a tip
+        self.assertCountEqual(self.manager.tx_storage.get_mempool_tips_index(), set())
+
+    def test_choose_tips(self):
+        genesis = self.manager.tx_storage.get_all_genesis()
+        genesis_txs_hashes = [tx.hash for tx in genesis if not tx.is_block]
+
+        b = add_new_block(self.manager, advance_clock=1)
+        # The txs parents are the genesis
+        self.assertCountEqual(set(b.parents[1:]), set(genesis_txs_hashes))
+        reward_blocks = add_blocks_unlock_reward(self.manager)
+        # No tips
+        self.assertEqual(len(self.manager.tx_storage.get_mempool_tips_index()), 0)
+
+        tx1 = add_new_transactions(self.manager, 1, advance_clock=1)[0]
+        # The tx parents will be the genesis txs still
+        self.assertCountEqual(set(tx1.parents), set(genesis_txs_hashes))
+        # The new tx will be a tip
+        self.assertCountEqual(self.manager.tx_storage.get_mempool_tips_index(), set([tx1.hash]))
+
+        tx2 = add_new_transactions(self.manager, 1, advance_clock=1)[0]
+        # The tx2 parents will be the tx1 and one of the genesis
+        self.assertTrue(tx1.hash in tx2.parents)
+        # The other parent will be one of tx1 parents
+        self.assertTrue(set(tx2.parents).issubset(set([tx1.hash] + tx1.parents)))
+        self.assertCountEqual(self.manager.tx_storage.get_mempool_tips_index(), set([tx2.hash]))
+
+        tx3 = add_new_transactions(self.manager, 1, advance_clock=1)[0]
+        # tx3 parents will be tx2 and one of tx2 parents
+        self.assertTrue(tx2.hash in tx3.parents)
+        self.assertTrue(set(tx3.parents).issubset(set([tx2.hash] + tx2.parents)))
+        self.assertCountEqual(self.manager.tx_storage.get_mempool_tips_index(), set([tx3.hash]))
+
+        b2 = add_new_block(self.manager, advance_clock=1)
+        # With new block there are no tips and block parents
+        # will be tx3 and one of tx3 parents
+        self.assertEqual(len(self.manager.tx_storage.get_mempool_tips_index()), 0)
+        self.assertTrue(tx3.hash in b2.parents)
+        self.assertTrue(reward_blocks[-1].hash in b2.parents)
+        self.assertTrue(set(b2.parents).issubset(set([tx3.hash] + [reward_blocks[-1].hash] + tx3.parents)))
+
+        tx4 = add_new_transactions(self.manager, 1, advance_clock=1)[0]
+        # tx4 had no tip, so the parents will be the last block parents
+        self.assertCountEqual(set(tx4.parents), set(b2.parents[1:]))
+        # Then tx4 will become a tip
+        self.assertCountEqual(self.manager.tx_storage.get_mempool_tips_index(), set([tx4.hash]))
+
+    def test_tips_twin(self):
+        # b = add_new_blocks(self.manager, 6, advance_clock=1)
```

Comment:
Should we remove it and the line below?

Reply:
They are commented out in the `test_tips_winner()` method as well. I think the variables are not used anymore and we can remove it. What do you think?

Reply:
Agree

Reply:
OK, removing...

Reply:
Done.